blockchain/indexers/manager.go

```diff
@@ -0,0 +1,648 @@
+		err := dropIndex(m.db, indexer.Key(), indexer.Name())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// maybeCreateIndexes determines if each of the enabled indexes have already
+// been created and creates them if not.
+func (m *Manager) maybeCreateIndexes(dbTx database.Tx) error {
+	indexesBucket := dbTx.Metadata().Bucket(indexTipsBucketName)
+	for _, indexer := range m.enabledIndexes {
+		// Nothing to do if the index tip already exists.
+		idxKey := indexer.Key()
+		if indexesBucket.Get(idxKey) != nil {
+			continue
+		}
+
+		// The tip for the index does not exist, so create it and
+		// invoke the create callback for the index so it can perform
+		// any one-time initialization it requires.
+		if err := indexer.Create(dbTx); err != nil {
+			return err
+		}
+
+		// Set the tip for the index to values which represent an
+		// uninitialized index.
+		err := dbPutIndexerTip(dbTx, idxKey, &wire.ShaHash{}, -1)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Init initializes the enabled indexes.  This is called during chain
+// initialization and primarily consists of catching up all indexes to the
+// current best chain tip.  This is necessary since each index can be disabled
+// and re-enabled at any time and attempting to catch-up indexes at the same
+// time new blocks are being downloaded would lead to an overall longer time to
+// catch up due to the I/O contention.
+//
+// This is part of the blockchain.IndexManager interface.
+func (m *Manager) Init(chain *blockchain.BlockChain) error {
+	// Nothing to do when no indexes are enabled.
+	if len(m.enabledIndexes) == 0 {
+		return nil
+	}
+
+	// Finish and drops that were previously interrupted.
+	if err := m.maybeFinishDrops(); err != nil {
+		return err
+	}
+
+	// Create the initial state for the indexes as needed.
+	err := m.db.Update(func(dbTx database.Tx) error {
+		// Create the bucket for the current tips as needed.
+		meta := dbTx.Metadata()
+		_, err := meta.CreateBucketIfNotExists(indexTipsBucketName)
+		if err != nil {
+			return err
+		}
+
+		return m.maybeCreateIndexes(dbTx)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Initialize each of the enabled indexes.
+	for _, indexer := range m.enabledIndexes {
+		if err := indexer.Init(); err != nil {
+			return err
+		}
+	}
+
+	// Rollback indexes to the main chain if their tip is an orphaned fork.
+	// This is fairly unlikely, but it can happen if the chain is
+	// reorganized while the index is disabled.  This has to be done in
+	// reverse order because indexes can depend on others.
+	for i := len(m.enabledIndexes); i > 0; i-- {
+		indexer := m.enabledIndexes[i-1]
+
+		// Fetch the current tip for the index.
+		var height int32
+		var hash *wire.ShaHash
+		err := m.db.View(func(dbTx database.Tx) error {
+			idxKey := indexer.Key()
+			hash, height, err = dbFetchIndexerTip(dbTx, idxKey)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
```

Comment:
Just ran into an issue while testing.

Steps to reproduce: 
  1. Start with a fresh `data` directory (no indexes, no blocks, etc)
  2. Sync up full chain
  3. Start with `btcd --txindex` or `btcd --addrindex`
 
Result: 
```
21:19:18 2016-03-29 [WRN] BTCD: open /home/laolu32/.btcd/btcd.conf: no such file or directory
21:19:18 2016-03-29 [INF] BTCD: Version 0.12.0-beta
21:19:18 2016-03-29 [INF] BTCD: Loading block database from '/home/laolu32/.btcd/data/testnet/blocks_ffldb'
21:19:18 2016-03-29 [INF] BTCD: Block database loaded
21:19:18 2016-03-29 [INF] INDX: Transaction index is enabled
21:19:18 2016-03-29 [ERR] BTCD: Unable to start server on [:18333]: block 0000000000000000000000000000000000000000000000000000000000000000 does not exist
```
`btcd` exits after printing the above log message.

Since none of the indexes has been created yet, `dbFetchIndexerTip` returns a `zeroHash` for the block hash. This hash isn't found in the main chain by `chain.MainChainHasBlock`, so the loop then proceeds to fetch a block with a block hash of all zeroes, which obviously does not exist. To remedy, there should be a check here for equality against the `zeroHash`, if `hash` is equivalent, then `continue` since the current index has never been created. Any other way to detect an uninitialized index should suffice also.